torch_xla/csrc/runtime/stablehlo_composite_helper.cc

```diff
@@ -0,0 +1,404 @@
+               builder.getF32FloatAttr(j.template get<float>())});
+          break;
+        case json::value_t::boolean:
+          named_attrs.push_back({builder.getStringAttr(key),
+                                 builder.getBoolAttr(j.template get<bool>())});
+          break;
+        case json::value_t::string:
+          named_attrs.push_back(
+              {builder.getStringAttr(key),
+               builder.getStringAttr(j.template get<std::string>())});
+          break;
+        default:
+          // Ignored unrecognizable attr json
+          break;
+      }
+    }
+    return builder.getDictionaryAttr(named_attrs);
+  }
+
+  void BuildStableHLOCompositeOp(
+      mlir::Operation* op,
+      const llvm::DenseMap<const mlir::Operation*, size_t>& op_line_num) {
+    mlir::ModuleOp module_op = getOperation();
+    mlir::MLIRContext* context = &getContext();
+    mlir::OpBuilder builder(context);
+
+    std::unique_ptr<BoundaryMetadata> metadata = GetBoundaryMetadata(op);
+    if (metadata == nullptr || metadata->is_input) {
+      return;
+    }
+    const auto& output_metadata = *metadata;
+
+    llvm::SetVector<mlir::Operation*> scope_ops_setvec;
+    llvm::SetVector<std::pair<mlir::Value, int64_t>> arg_pos_setvec;
+    llvm::SmallVector<mlir::Operation*> processing({op});
+
+    // Reverse graph traversal: from boundary output op to boundary input op,
+    // global function arg, or stablehlo constant.
+    while (!processing.empty()) {
+      mlir::Operation* curr_op = processing.back();
+      processing.pop_back();
+      if (scope_ops_setvec.contains(curr_op)) {
+        continue;
+      }
+
+      if (auto curr_metadata_ptr = GetBoundaryMetadata(curr_op);
+          curr_metadata_ptr != nullptr) {
+        const auto& curr_metadata = *curr_metadata_ptr;
+        if (curr_metadata.is_input &&
+            curr_metadata.boundary_id() == output_metadata.boundary_id()) {
+          // Terminal condition: boundary input op.
+          arg_pos_setvec.insert({curr_op->getResult(0).dyn_cast<mlir::Value>(),
+                                 curr_metadata.pos});
+          continue;
+        }
+      }
+
+      scope_ops_setvec.insert(curr_op);
+      for (mlir::Value value : curr_op->getOperands()) {
+        mlir::Operation* def_op = value.getDefiningOp();
+        if (def_op == nullptr) {
+          // Terminal condition: Global function arg
+          arg_pos_setvec.insert({value, std::numeric_limits<int64_t>::max()});
+        } else if (llvm::isa<mlir::stablehlo::ConstantOp>(op)) {
+          // Terminal condition: constant
+          scope_ops_setvec.insert(def_op);
+        } else {
+          processing.push_back(def_op);
+        }
+      }
+    }
+    // Sorts all ops within the boundary by their line numbers in the input
+    // MLIR. The ops will be duplicated to the impl function following this
+    // order.
+    auto scope_ops = scope_ops_setvec.takeVector();
+    for (auto& op : scope_ops) {
```

Comment:
This sounds fragile.. is it needed for correctness? I.e. a program could be all be a single line? What would break if it were?

Reply:
Yes. Otherwise it'd be very hard to determine the insertion order.

It's actually not the "line number" in the input text but the order when iterating through ops in the function in MLIR. So it's guaranteed that each op will have a unique "line number"

Reply:
Ah, that makes more sense. I'll dig around if theres a more convenient way to do this in MLIR. Can we use a different name than `line_number`? `topological_position`, `encounter_position`, etc?

Reply:
Renamed the lut to `op_order_map`